File: ch04/switchbtn/switchbtn_ex4.c
#include <stdio.h>
#include <pigpiod_if2.h> 
#define PINNO 23
int count;
void cb_falling_func(int pi, unsigned user_gpio, unsigned level, uint32_t tick)
{
//    printf("cb_falling_func gpio:%d count:%d\n", user_gpio, count);
}

void cb_rising_func(int pi, unsigned user_gpio, unsigned level, uint32_t tick)
{
//    printf("cb_rising_func gpio:%d count:%d\n", user_gpio, count);
    count++;
}

int main()
{
    int pi;
    if((pi = pigpio_start(NULL, NULL)) < 0){
        fprintf(stderr, "%s\n", pigpio_error(pi));  return 1;
    }

    set_mode(pi, PINNO, PI_INPUT);
    set_pull_up_down(pi, PINNO, PI_PUD_UP);
    callback(pi, PINNO, FALLING_EDGE, cb_falling_func);
    callback(pi, PINNO, RISING_EDGE, cb_rising_func);

    getchar();
    printf("버튼 누른 총 횟수:%d\n", count);

    pigpio_stop(pi);
    return 0;
}

File: ch04/switchbtn/switchbtn_ex4.py
#!/usr/bin/python3
import pigpio
import time

count = 0
def cbf_falling(gpio, level, tick):
    print("gpio:", gpio, " level:", level, " tick:", tick, "\n")

def cbf_rising(gpio, level, tick):
    global count
    count += 1
    print("count:", count)
    print("gpio:", gpio, " level:", level, " tick:", tick, "\n")

GPIO = 23

pi = pigpio.pi()
if not pi.connected:
    exit()

pi.set_mode(GPIO, pigpio.INPUT)
pi.set_pull_up_down(GPIO, pigpio.PUD_UP)

pi.callback(GPIO, pigpio.FALLING_EDGE, cbf_falling)
pi.callback(GPIO, pigpio.RISING_EDGE, cbf_rising)
input()
print("버튼 누른 총 횟수:", count)

pi.stop()
